File: client/tws_foundation_client/lib/src/services/business/drivers/driver.dart
import 'package:csm_client/csm_client.dart';
import 'package:tws_foundation_client/tws_foundation_client.dart';

/// {entity} class.
///
///
final class Driver extends EntityB<Driver> {
  /// [Driver.employee] property key for [DataMap].
  static const String kEmployee = "employee";

  /// [Driver.driverType] property key for [DataMap].
  static const String kDriverType = "driverType";

  /// [Driver.licenseExpiration] property key for [DataMap].
  static const String kLicenseExpiration = "licenseExpiration";

  /// [Driver.drugAlcRegistrationDate] property key for [DataMap].
  static const String kDrugalcRegistrationDate = "drugalcRegistrationDate";

  /// [Driver.pullNoticeRegistrationDate] property key for [DataMap].
  static const String kPullnoticeRegistrationDate = "pullnoticeRegistrationDate";

  /// [Driver.twic] property key for [DataMap].
  static const String kTwic = "twic";

  /// [Driver.twicExpiration] property key for [DataMap].
  static const String kTwicExpiration = "twicExpiration";

  /// [Driver.visa] property key for [DataMap].
  static const String kVisa = "visa";

  /// [Driver.visaExpiration] property key for [DataMap].
  static const String kVisaExpiration = "visaExpiration";

  /// [Driver.fast] property key for [DataMap].
  static const String kFast = "fast";

  /// [Driver.fastExpiration] property key for [DataMap].
  static const String kFastExpiration = "fastExpiration";

  /// [Driver.anam] property key for [DataMap].
  static const String kAnam = "anam";

  /// [Driver.anamExpiration] property key for [DataMap].
  static const String kAnamExpiration = "anamExpiration";

  //! --> Properties

  /// "Fast" permit number.
  ///
  /// Length must be 24
  String? fast;

  /// TODO: Define.
  ///
  /// Length must be 24
  String? anam;

  /// USA visa document number.
  ///
  /// Length must be 24.
  String? visa;

  /// TODO: Define.
  ///
  /// Length must be 12.
  String? twic;

  /// TODO: Define.
  ///
  /// Length must be 12.
  String? driverType;

  /// Driver's license expiration date.
  DateTime? licenseExpiration;

  /// TODO: Define.
  DateTime? drugAlcRegistrationDate;

  /// TODO: Define.
  DateTime? pullNoticeRegistrationDate;

  /// TODO: Define.
  DateTime? twicExpiration;

  /// USA visa expiration date.
  DateTime? visaExpiration;

  /// "fast" permit expiration date.
  DateTime? fastExpiration;

  /// TODO: Define.
  DateTime? anamExpiration;

  //! <-- Properties

  //! --> Relations

  /// [Employee] information.
  Employee employee = Employee();

  //! <-- Relations

  /// Creates a new [Driver] instance.
  Driver();

  @override
  void decode(DataMap encode) {
    fast = encode.get(kFast);
    anam = encode.get(kAnam);
    visa = encode.get(kVisa);
    twic = encode.get(kTwic);
    driverType = encode.get(kDriverType);
    licenseExpiration = encode.get(kLicenseExpiration);
    drugAlcRegistrationDate = encode.get(kDrugalcRegistrationDate);
    pullNoticeRegistrationDate = encode.get(kPullnoticeRegistrationDate);
    twicExpiration = encode.get(kTwicExpiration);
    visaExpiration = encode.get(kVisaExpiration);
    fastExpiration = encode.get(kFastExpiration);
    anamExpiration = encode.get(kAnamExpiration);

    employee = encode.getEntity(() => Employee(), kEmployee) ?? employee;

    super.decode(encode);
  }

  @override
  DataMap encode([DataMap? entityObject]) {
    return super.encode(
      <String, Object?>{
        kFast: fast,
        kAnam: anam,
        kVisa: visa,
        kTwic: twic,
        kDriverType: driverType,
        kLicenseExpiration: licenseExpiration?.toIso8601String(),
        kDrugalcRegistrationDate: drugAlcRegistrationDate?.toIso8601String(),
        kPullnoticeRegistrationDate: pullNoticeRegistrationDate?.toIso8601String(),
        kTwicExpiration: twicExpiration?.toIso8601String(),
        kVisaExpiration: visaExpiration?.toIso8601String(),
        kFastExpiration: fastExpiration?.toIso8601String(),
        kAnamExpiration: anamExpiration?.toIso8601String(),
        kEmployee: employee.encode(),
      },
    );
  }

  @override
  List<EntityInvalidation<Driver>> evaluate() {
    return <EntityInvalidation<Driver>>[];
  }
}

File: server/TWS Foundation/TWS Business/Entities/Drivers/Driver.cs
using System.ComponentModel.DataAnnotations;

using CSM_Database_Core.Core.Attributes;
using CSM_Database_Core.Core.Extensions;

using Microsoft.EntityFrameworkCore.Metadata.Builders;

using TWS_Business.Bases;
using TWS_Business.Entities.Employees;

namespace TWS_Business.Entities.Drivers;

/// <summary>
///     [Entity] that represents an internal business driver (Tuck operator).
/// </summary>
public class Driver
    : BCommonScopeEntity<Driver_Common> {

    #region Properties

    /// <summary>
    ///     Fast permit number.
    /// </summary>
    [StringLength(12, MinimumLength = 12)]
    public string? Fast { get; set; }

    /// <summary>
    ///     TBD
    /// </summary>
    [StringLength(24, MinimumLength = 24)]
    public string? ANAM { get; set; }

    /// <summary>
    ///     USA Visa document number.
    /// </summary>
    [StringLength(12, MinimumLength = 12)]
    public string? VISA { get; set; }

    /// <summary>
    ///     TBD
    /// </summary>
    [StringLength(12, MinimumLength = 12)]
    public string? TWIC { get; set; }

    /// <summary>
    ///     Driver type name.
    /// </summary>
    [StringLength(12, MinimumLength = 1)]
    public string? DriverType { get; set; }

    /// <summary>
    ///     Driver licence expiration date.
    /// </summary>
    public DateOnly? LicenseExpiration { get; set; }

    /// <summary>
    ///     TBD
    /// </summary>
    public DateOnly? DrugalcRegistrationDate { get; set; }

    /// <summary>
    ///     TBD
    /// </summary>
    public DateOnly? PullnoticeRegistrationDate { get; set; }

    /// <summary>
    ///     TDB
    /// </summary>
    public DateOnly? TwicExpiration { get; set; }

    /// <summary>
    ///     USA Visa expiration date.
    /// </summary>
    public DateOnly? VisaExpiration { get; set; }

    /// <summary>
    ///     Fast permit expiration date.
    /// </summary>
    public DateOnly? FastExpiration { get; set; }

    /// <summary>
    ///     TBD
    /// </summary>
    public DateOnly? AnamExpiration { get; set; }

    #endregion

    #region Relations

    /// <summary>
    ///     <see cref="Employees.Employee"/> information.
    /// </summary>
    [EntityRelation]
    public Employee Employee { get; set; } = default!;

    #endregion

    protected override void DesignScopeEntity(EntityTypeBuilder etBuilder) {
        etBuilder.Property(nameof(DriverType)).HasMaxLength(12);
        etBuilder.Property(nameof(TWIC)).HasMaxLength(12);
        etBuilder.Property(nameof(VISA)).HasMaxLength(12);
        etBuilder.Property(nameof(Fast)).HasMaxLength(12);
        etBuilder.Property(nameof(ANAM)).HasMaxLength(24);

        etBuilder.Link<Driver, Employee>(
                nameof(Employee),
                Required: true,
                Auto: true,
                Index: true
            );
    }
}
